cloudwatch-integ/CloudwatchMonitoring.cpp

```diff
@@ -0,0 +1,409 @@
+            return "Kilobits_Second";
+        default:
+            return "Unknown unit";
+    }
+}
+
+VOID CloudwatchMonitoring::push(const MetricDatum& datum)
+{
+    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+    MetricDatum single = datum;
+    MetricDatum aggregated = datum;
+
+    single.AddDimensions(this->channelDimension);
+    single.AddDimensions(this->labelDimension);
+    aggregated.AddDimensions(this->labelDimension);
+
+    cwRequest.SetNamespace(DEFAULT_CLOUDWATCH_NAMESPACE);
+    cwRequest.AddMetricData(single);
+    cwRequest.AddMetricData(aggregated);
+
+    auto asyncHandler = [this](const Aws::CloudWatch::CloudWatchClient* cwClient, const Aws::CloudWatch::Model::PutMetricDataRequest& request,
+                               const Aws::CloudWatch::Model::PutMetricDataOutcome& outcome,
+                               const std::shared_ptr<const Aws::Client::AsyncCallerContext>& context) {
+        UNUSED_PARAM(cwClient);
+        UNUSED_PARAM(request);
+        UNUSED_PARAM(context);
+
+        if (!outcome.IsSuccess()) {
+            DLOGE("Failed to put sample metric data: %s", outcome.GetError().GetMessage().c_str());
+        } else {
+            DLOGS("Successfully put sample metric data");
+        }
+        this->pendingMetrics--;
+    };
+    this->pendingMetrics++;
+    this->client.PutMetricDataAsync(cwRequest, asyncHandler);
+
+    std::stringstream ss;
+
+    ss << "Emitted the following metric:\n\n";
+    ss << "  Name       : " << datum.GetMetricName() << '\n';
+    ss << "  Unit       : " << unitToString(datum.GetUnit()) << '\n';
+
+    ss << "  Values     : ";
+    auto& values = datum.GetValues();
+    // If the datum uses single value, GetValues will be empty and the data will be accessible
+    // from GetValue
+    if (values.empty()) {
+        ss << datum.GetValue();
+    } else {
+        for (auto i = 0; i < values.size(); i++) {
+            ss << values[i];
+            if (i != values.size() - 1) {
+                ss << ", ";
+            }
+        }
+    }
+    ss << '\n';
+
+    ss << "  Dimensions : ";
+    auto& dimensions = datum.GetDimensions();
+    if (dimensions.empty()) {
+        ss << "N/A";
+    } else {
+        ss << '\n';
+        for (auto& dimension : dimensions) {
+            ss << "    - " << dimension.GetName() << "\t: " << dimension.GetValue() << '\n';
+        }
+    }
+    ss << '\n';
+
+    DLOGD("%s", ss.str().c_str());
+}
+
+VOID CloudwatchMonitoring::pushExitStatus(STATUS retStatus)
+{
+    MetricDatum datum;
+    Dimension statusDimension;
+    CHAR status[MAX_STATUS_CODE_LENGTH];
+
+    statusDimension.SetName("Code");
+    SPRINTF(status, "0x%08x", retStatus);
+    statusDimension.SetValue(status);
+
+    datum.SetMetricName("ExitStatus");
+    datum.SetValue(1.0);
+    datum.SetUnit(Aws::CloudWatch::Model::StandardUnit::Count);
+
+    datum.AddDimensions(statusDimension);
+
+    this->push(datum);
+}
+
```

Comment:
Is there a reason `VOID CloudwatchMonitoring::pushSignalingRoundtripStatus(STATUS retStatus)`, `VOID CloudwatchMonitoring::pushSignalingRoundtripLatency(UINT64 delay, Aws::CloudWatch::Model::StandardUnit unit)`, `VOID CloudwatchMonitoring::pushSignalingConnectionDuration(UINT64 duration, Aws::CloudWatch::Model::StandardUnit unit)`, `VOID CloudwatchMonitoring::pushSignalingInitDelay(UINT64 delay, Aws::CloudWatch::Model::StandardUnit unit)`, `VOID CloudwatchMonitoring::pushICEHolePunchingDelay(UINT64 delay, Aws::CloudWatch::Model::StandardUnit unit)` are not included here but are in the canary?

Reply:
There isnt going to be a signaling specific canary. These are integrated in profiling metrics now.